src/compiler/binder.ts

```diff
@@ -278,9 +278,11 @@ namespace ts {
         function declareSymbol(symbolTable: SymbolTable, parent: Symbol, node: Declaration, includes: SymbolFlags, excludes: SymbolFlags): Symbol {
             Debug.assert(!hasDynamicName(node));

```

Comment:
Consider adding a `Debug.assert` that if `node` is a `BinaryExpression`, you're also in a JS file.